### ISSUE TITLE
`C_Cpp.clang_format_formatOnSave` has a delay for about 1 second.

### ISSUE THREAD

**Issue by be5invis** [microsoft/vscode-cpptools#184]
Windows 10 x64, clang/LLVM 3.8.1

**Comment by sean-mcmanus** [comment]
I need more repro details. How large is your file? Does the delay occur when doing manual formatting? Seems like this may be expected. clang-format itself takes time and we do processing internally too. Do you want us to try to optimize it so it's faster?

**Comment by be5invis** [comment]
@sean-mcmanus
1. code is less then 100 lines. like this
   
   ``` c++
   #include "base64.h"
   static const uint8_t base64_table[64] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
   
   uint8_t *base64_encode(const uint8_t *src, size_t len, size_t *out_len) {
   uint8_t *out, *pos;
   const uint8_t *end, *in;
   size_t olen;
   
   olen = len * 4 / 3 + 4; /* 3-byte blocks to 4-byte */
   olen++;                 /* nul termination */
   out = malloc(sizeof(uint8_t) * olen);
   if (out == NULL) return NULL;
   
   end = src + len;
   in = src;
   pos = out;
   while (end - in >= 3) {
       *pos++ = base64_table[in[0] >> 2];
       *pos++ = base64_table[((in[0] & 0x03) << 4) | (in[1] >> 4)];
       *pos++ = base64_table[((in[1] & 0x0f) << 2) | (in[2] >> 6)];
       *pos++ = base64_table[in[2] & 0x3f];
       in += 3;
   }
   
   if (end - in) {
       *pos++ = base64_table[in[0] >> 2];
       if (end - in == 1) {
           *pos++ = base64_table[(in[0] & 0x03) << 4];
           *pos++ = '=';
       } else {
           *pos++ = base64_table[((in[0] & 0x03) << 4) | (in[1] >> 4)];
           *pos++ = base64_table[(in[1] & 0x0f) << 2];
       }
       *pos++ = '=';
   }
   
   *pos = '\0';
   if (out_len) *out_len = pos - out;
   return out;
   }
   
   uint8_t *base64_decode(const uint8_t *src, size_t len, size_t *out_len) {
   uint8_t dtable[256], *out, *pos, in[4], block[4], tmp;
   size_t i, count;
   
   memset(dtable, 0x80, 256);
   for (i = 0; i < sizeof(base64_table); i++)
       dtable[base64_table[i]] = i;
   dtable['='] = 0;
   
   count = 0;
   for (i = 0; i < len; i++) {
       if (dtable[src[i]] != 0x80) count++;
   }
   
   if (count % 4) return NULL;
   
   pos = out = malloc(sizeof(uint8_t) * count);
   if (out == NULL) return NULL;
   
   count = 0;
   for (i = 0; i < len; i++) {
       tmp = dtable[src[i]];
       if (tmp == 0x80) continue;
   
       in[count] = src[i];
       block[count] = tmp;
       count++;
       if (count == 4) {
           *pos++ = (block[0] << 2) | (block[1] >> 4);
           *pos++ = (block[1] << 4) | (block[2] >> 2);
           *pos++ = (block[2] << 6) | block[3];
           count = 0;
       }
   }
   
   if (pos > out) {
       if (in[2] == '=')
           pos -= 2;
       else if (in[3] == '=')
           pos--;
   }
   
   *out_len = pos - out;
   return out;
   }
   ```
2. No delay when manual formatting.
3. No delay for `clang-format` extension’s format-on-save.

**Comment by sean-mcmanus** [comment]
Thanks for the repro info. Unfortunately, I'm not able to repro this on my Windows or Linux machines. The only major difference I see between formatOnSave and manually formatting is that we do an additional save operation after the formatting is done. So it seems like something strange is going on in your repro. Maybe another VS Code extension or some virus scan process is causing a delay?

**Comment by be5invis** [comment]
@sean-mcmanus 
I've uninstalled all the extensions except for `C/C++`, and the delay is still there.
I am using a `.clang-format` file.

**Comment by sean-mcmanus** [comment]
I can't repro it with the .clang-format file I'm using. Can you repro it without the .clang-format file (i.e. using the default "Visual Studio" style we added)? Otherwise, can you attach the .clang-format file you're using.

**Comment by be5invis** [comment]
@sean-mcmanus Switching style to “Visual Studio” still has a delay.
I’ve installed Clang/LLVM manually. Is it related?

**Comment by be5invis** [comment]
Wait, I've tried a folder with only one .cpp file, and this time saving has no delay.
So it is a performance problem...
I am editing [this](https://github.com/caryll/otfcc), a project with many .c/.cpp files. I've set `dep/` and `lib/` into `includes` in `c_cpp_properties.json`.
It may be something else preventing format... I think.

**Comment by sean-mcmanus** [comment]
I downloaded that project and set the includePaths and I didn't see a problem, but when I created over 1000 C++ files in the folder I started to repro the bug. The slowness is actually caused by our "browsing" code that is making the 1st save slow (the format/save that happens afterwards is fast). We'll try to get a fix in a future release. Thanks for reporting this problem and helping to discover a repro.

**Comment by sean-mcmanus** [comment]
I have a fix for the next release. I skip our browse parsing on the 1st save and just do it after the formatting or 2nd save.

**Comment by sean-mcmanus** [comment]
The fix won't be in the upcoming 0.9.1 release, but it should be in the next one.

**Comment by sean-mcmanus** [comment]
Fixed in 0.9.2.
